### ISSUE TITLE
4.0.0 - docker web panic runtime error SIGSEGV

### ISSUE THREAD

**Issue by eedwards-sk** [concourse/concourse#2526]
# Bug Report

* Concourse version: 4.0.0
* Deployment type (BOSH/Docker/binary): Docker (web / db / linux worker) Darwin (worker)
* Infrastructure/IaaS:
* Browser (if applicable):
* Did this used to work? No

I'm trying to run a concourse stack locally with a CI build tool. The web container is ending up in a restart loop spammed with this in the logs. Seems to be uncaught exception of sorts?

```
goroutine 1 [running]:
crypto/rsa.(*PrivateKey).Public(0x0, 0x0, 0x0)
	/usr/local/go/src/crypto/rsa/rsa.go:92 +0x5
golang.org/x/crypto/ssh.NewSignerFromSigner(0x7f69ac53c470, 0xc420190408, 0xc420190408, 0x7f69ac53c470, 0xc420190408, 0xec4101)
	/tmp/build/9674af12/concourse/src/golang.org/x/crypto/ssh/keys.go:681 +0x35
golang.org/x/crypto/ssh.NewSignerFromKey(0x299b760, 0xc420190408, 0xec4178, 0x30, 0x2afd220, 0xc4206193f8)
	/tmp/build/9674af12/concourse/src/golang.org/x/crypto/ssh/keys.go:656 +0x163
github.com/concourse/tsa/tsacmd.(*TSACommand).configureSSHServer(0xc4201809c0, 0xc420387230, 0xc420591bb0, 0x1, 0x1, 0x0, 0x0, 0x0, 0xc422d46f00, 0xc4206caca0, ...)
	/tmp/build/9674af12/concourse/src/github.com/concourse/tsa/tsacmd/command.go:203 +0x18e
github.com/concourse/tsa/tsacmd.(*TSACommand).Runner(0xc4201809c0, 0xc420591890, 0x0, 0x1, 0x143ecd40, 0xc42083f360, 0xc4204aa800, 0x0)
	/tmp/build/9674af12/concourse/src/github.com/concourse/tsa/tsacmd/command.go:101 +0x249
main.(*WebCommand).Runner(0xc4207fa708, 0xc420591890, 0x0, 0x1, 0x2149481, 0xc4202e0100, 0xc42079d660, 0x21499d0)
	/tmp/build/9674af12/concourse/src/github.com/concourse/bin/cmd/concourse/web.go:67 +0x118
main.(*WebCommand).Execute(0xc4207fa708, 0xc420591890, 0x0, 0x1, 0x28b59a0, 0x2b52f80)
	/tmp/build/9674af12/concourse/src/github.com/concourse/bin/cmd/concourse/web.go:42 +0x4d
github.com/vito/twentythousandtonnesofcrudeoil.installEnv.func2(0x7f69ac5867c0, 0xc4207fa708, 0xc420591890, 0x0, 0x1, 0x1, 0xc42079d7a8)
	/tmp/build/9674af12/concourse/src/github.com/vito/twentythousandtonnesofcrudeoil/environment.go:40 +0x8a
github.com/jessevdk/go-flags.(*Parser).ParseArgs(0xc42029db00, 0xc4200d6010, 0x1, 0x1, 0xc42079d840, 0x21492df, 0xc42029db00, 0x2bb0ae6, 0xa)
	/tmp/build/9674af12/concourse/src/github.com/jessevdk/go-flags/parser.go:314 +0x7c7
github.com/jessevdk/go-flags.(*Parser).Parse(0xc42029db00, 0x2bb0ae6, 0xa, 0xc4202e1c80, 0x0, 0x0)
	/tmp/build/9674af12/concourse/src/github.com/jessevdk/go-flags/parser.go:186 +0x71
main.main()
	/tmp/build/9674af12/concourse/src/github.com/concourse/bin/cmd/concourse/main.go:33 +0x1a4
{"timestamp":"1535053975.540039062","source":"atc","message":"atc.dex.event","log_level":1,"data":{"fields":{},"message":"keys expired, rotating","session":"5"}}
{"timestamp":"1535053975.848279238","source":"atc","message":"atc.dex.event","log_level":1,"data":{"fields":{},"message":"keys rotated, next rotation: 2018-08-24 01:52:55.8481602 +0000 UTC m=+21601.550173301","session":"5"}}
{"timestamp":"1535053975.888875246","source":"atc","message":"atc.dex.event","log_level":1,"data":{"fields":{},"message":"keys expired, rotating","session":"25"}}
{"timestamp":"1535053975.981125832","source":"atc","message":"atc.dex.event","log_level":1,"data":{"fields":{},"message":"keys rotated, next rotation: 2018-08-24 01:52:55.9810883 +0000 UTC m=+21601.683101401","session":"25"}}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xffc135]

goroutine 1 [running]:
crypto/rsa.(*PrivateKey).Public(0x0, 0x0, 0x0)
	/usr/local/go/src/crypto/rsa/rsa.go:92 +0x5
golang.org/x/crypto/ssh.NewSignerFromSigner(0x7f6c811b3a18, 0xc42000e7b8, 0xc42000e7b8, 0x7f6c811b3a18, 0xc42000e7b8, 0xec4101)
	/tmp/build/9674af12/concourse/src/golang.org/x/crypto/ssh/keys.go:681 +0x35
golang.org/x/crypto/ssh.NewSignerFromKey(0x299b760, 0xc42000e7b8, 0xec4178, 0x30, 0x2afd220, 0xc4227e13f8)
	/tmp/build/9674af12/concourse/src/golang.org/x/crypto/ssh/keys.go:656 +0x163
github.com/concourse/tsa/tsacmd.(*TSACommand).configureSSHServer(0xc4201aa3c0, 0xc4204d09e0, 0xc4204b0370, 0x1, 0x1, 0x0, 0x0, 0x0, 0xc42016d500, 0xc420666740, ...)
	/tmp/build/9674af12/concourse/src/github.com/concourse/tsa/tsacmd/command.go:203 +0x18e
github.com/concourse/tsa/tsacmd.(*TSACommand).Runner(0xc4201aa3c0, 0xc4204b0070, 0x0, 0x1, 0x143ecd40, 0xc421962de0, 0xc420727800, 0x0)
	/tmp/build/9674af12/concourse/src/github.com/concourse/tsa/tsacmd/command.go:101 +0x249
main.(*WebCommand).Runner(0xc4205e5508, 0xc4204b0070, 0x0, 0x1, 0x2149481, 0xc4203b8e00, 0xc42062b660, 0x21499d0)
	/tmp/build/9674af12/concourse/src/github.com/concourse/bin/cmd/concourse/web.go:67 +0x118
main.(*WebCommand).Execute(0xc4205e5508, 0xc4204b0070, 0x0, 0x1, 0x28b59a0, 0x2b52f80)
	/tmp/build/9674af12/concourse/src/github.com/concourse/bin/cmd/concourse/web.go:42 +0x4d
github.com/vito/twentythousandtonnesofcrudeoil.installEnv.func2(0x7f6c81204950, 0xc4205e5508, 0xc4204b0070, 0x0, 0x1, 0x1, 0xc42062b7a8)
	/tmp/build/9674af12/concourse/src/github.com/vito/twentythousandtonnesofcrudeoil/environment.go:40 +0x8a
github.com/jessevdk/go-flags.(*Parser).ParseArgs(0xc4202b7620, 0xc42000c090, 0x1, 0x1, 0xc42062b840, 0x21492df, 0xc4202b7620, 0x2bb0ae6, 0xa)
	/tmp/build/9674af12/concourse/src/github.com/jessevdk/go-flags/parser.go:314 +0x7c7
github.com/jessevdk/go-flags.(*Parser).Parse(0xc4202b7620, 0x2bb0ae6, 0xa, 0xc4203b9980, 0x0, 0x0)
	/tmp/build/9674af12/concourse/src/github.com/jessevdk/go-flags/parser.go:186 +0x71
main.main()
	/tmp/build/9674af12/concourse/src/github.com/concourse/bin/cmd/concourse/main.go
```

**Comment by eedwards-sk** [comment]
Looks like a permissions issue on the keys. I'm diagnosing it at the system level now, so it's not a concourse problem, but I still suspect the error message could be improved (like actually handling this exception).

**Comment by cperry169** [comment]
Did you find a solution to the error message. If so please send me an email to cperry173@outlook.com.  This is blocking the upgrade from 3.9.0 to 4.1.0

**Comment by eedwards-sk** [comment]
This is still an issue. We absolutely need clear error messages for this.

I'm now losing hours over chasing this problem.

Please help.

#2135 #2512

**Comment by eedwards-sk** [comment]
tracing down the issue

https://github.com/concourse/concourse/blob/master/tsa/tsacmd/command.go#L93

```
config, err := cmd.configureSSHServer(sessionAuthTeam, cmd.AuthorizedKeys.Keys, teamAuthorizedKeys)
```

https://github.com/concourse/tsa/blob/master/tsacmd/command.go#L197

```
signer, err := ssh.NewSignerFromKey(cmd.HostKey)
```

**Comment by eedwards-sk** [comment]
I'm working on setting up a debug environment with v4.2.2 to see if I can track down the source.

**Comment by eedwards-sk** [comment]
1. almost impossible to debug concourse right now on v4.2.2 due to bug in golang that won't be fixed until 1.12. Basically can't debug plugins. Sigh.

2. I reproduced the error on my main install after destroying my keys. This appears to be a cause of a bad key issue getting swallowed with no proper error.

https://github.com/concourse/docs/pull/124

has some info, I'm still working on finding the solve.

**Comment by eedwards-sk** [comment]
ok, workaround found @cperry169 

followed instructions in https://github.com/concourse/concourse/issues/2592#issuecomment-435193365 to generate a compatible key

so e.g.

```
openssl genpkey -algorithm RSA -out session_signing_key -pkeyopt rsa_keygen_bits:2048
```

followed by

```
chmod 0400 session_signing_key
```

and finally

```
ssh-keygen -f session_signing_key -y > session_signing_key.pub
```
